### PR TITLE
Add Device Properties Functionality

### PR DESCRIPTION
**Pull request by jillcary**
# Type of change? 

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 
This is the merge of the following PRs and branches for re-adding device properties support back into the latest simulation branch. 

- [Add Device Properties logic to internal simulation](https://github.com/Azure/device-simulation-dotnet/pull/173)
- [Add DevicePropertiesActor and push reported properties updates to the IoT Hub](https://github.com/Azure/device-simulation-dotnet/pull/181)
- [Register for Desired Property State Changes](https://github.com/Azure/device-simulation-dotnet/pull/184)
- [Firmware update method calls updateProperty callback with all parameters](https://github.com/Azure/device-simulation-dotnet/pull/186)
- [Add Methods](https://github.com/Azure/device-simulation-dotnet/pull/182) 
- [Parvezp/add methods](https://github.com/Azure/device-simulation-dotnet/pull/182)

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [x] The change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: Services/Models/SmartDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            }
+            else
+            {
+                this.dictionary.Add(key, value);
+            }
+
+            this.Changed = true;
+        }
+
+        public void SetAll(Dictionary<string, object> newState)
```
Comment (dluc) [inline Services/Models/SmartDictionary.cs]: I'd force this to be `ConcurrentDictionary`

Comment (jillcary) [reply Services/Models/SmartDictionary.cs]: Done.

File: SimulationAgent/DeviceProperties/DevicePropertiesActor.cs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.Exceptions;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.DeviceProperties
+{
+    public interface IDevicePropertiesActor
+    {
+        ISmartDictionary DeviceProperties { get; }
+        ISmartDictionary DeviceState { get; }
+        IDeviceClient Client { get; }
+        Device Device { get; }
```
Comment (dluc) [inline SimulationAgent/DeviceProperties/DevicePropertiesActor.cs]: looks like this is not used, can we remove it?

File: SimulationAgent/DeviceProperties/DevicePropertiesActor.cs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+        /// <summary>
+        /// Azure IoT Hub client
+        /// </summary>
+        public IDeviceClient Client => this.deviceConnectionActor.Client;
+
+        /// <summary>
+        /// Azure IoT Hub Device instance
+        /// </summary>
+        public Device Device => this.deviceConnectionActor.Device;
```
Comment (dluc) [inline SimulationAgent/DeviceProperties/DevicePropertiesActor.cs]: looks like this is not used, can we remove it?

File: Services/DevicePropertiesRequest.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services
+{
+    public interface IDevicePropertiesRequest
+    {
+        Task RegisterDevicePropertyUpdatesAsync(
+            string deviceId,
+            ISmartDictionary deviceProperties);
+
+        Task OnPropertyUpdateRequestedCallback(
```
Comment (dluc) [inline Services/DevicePropertiesRequest.cs]: this method is not used outside the class below, can we then remove it from the interface and make it private?

File: Services/DevicePropertiesRequest.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            this.log.Debug("Setting up callback for desired properties updates.", () => new { this.deviceId });
+
+            // Set callback that IoT Hub calls whenever the client receives a desired properties state update.
+            await this.client.SetDesiredPropertyUpdateCallbackAsync(OnPropertyUpdateRequestedCallback, null);
+
+            this.log.Debug("Callback for desired properties updates setup successfully", () => new { this.deviceId });
+        }
+
+        /// <summary>
+        /// When a desired property change is requested, update the internal the device state properties
```
Comment (dluc) [inline Services/DevicePropertiesRequest.cs]: typo

File: Services/DevicePropertiesRequest.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                    // Only update if key doesn't exist or value has changed 
+                    if (!this.deviceProperties.Has(item.Key) ||
+                        (item.Value.ToString() != this.deviceProperties.Get(item.Key).ToString()))
+                    {
+                        // Update existing property or create new property if key doesn't exist.
+                        this.deviceProperties.Set(item.Key, item.Value);
+                    }
+                }
+            }
+            catch (Exception e)
```
Comment (dluc) [inline Services/DevicePropertiesRequest.cs]: when could this happen?

File: Services/DevicePropertiesRequest.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                this.log.Error("Error updating internal device state properties", () => new { e, this.deviceId, desiredProperties });
+            }
+
+            this.log.Debug("Desired property update successfully reported to internal state", () => new { this.deviceId, desiredProperties });
```
Comment (dluc) [inline Services/DevicePropertiesRequest.cs]: could you rephrase? something like "device properties updated to the desired values"

File: Services/DevicePropertiesRequest.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        public DevicePropertiesRequest(Azure.Devices.Client.DeviceClient client, ILogger logger)
+        {
+            this.client = client;
+            this.log = logger;
+            this.deviceId = string.Empty;
+        }
+
+        public async Task RegisterDevicePropertyUpdatesAsync(string deviceId, ISmartDictionary deviceProperties)
+        {
+            if (this.deviceId != string.Empty)
```
Comment (dluc) [inline Services/DevicePropertiesRequest.cs]: I would pass the deviceId to the ctor class, and have a "registered" boolean for this purpose

File: Services/DevicePropertiesRequest.cs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services
+{
+    public interface IDevicePropertiesRequest
```
Comment (dluc) [inline Services/DevicePropertiesRequest.cs]: the name is a bit strange, e.g. why "request" ? I would have chosen something like 
```
IDeviceProperties {
    Task OnChangeUpdateAsync(ISmartDictionary deviceProperties)
}
```

File: Services/DevicePropertiesRequest.cs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            // This is where custom code for handling specific desired property changes could be added.
+            // For the purposes of the simulation service, we have chosen to write the desired properties
+            // directly to the reported properties. 
+            try
+            {
+                foreach (KeyValuePair<string, object> item in desiredProperties)
+                {
+                    // Only update if key doesn't exist or value has changed 
+                    if (!this.deviceProperties.Has(item.Key) ||
+                        (item.Value.ToString() != this.deviceProperties.Get(item.Key).ToString()))
```
Comment (dluc) [inline Services/DevicePropertiesRequest.cs]: wondering if `ToString` is the right way of comparing values. For instance, what if the value is a complex object? Does `ToString` serialize the actual value?

Comment (jillcary) [reply Services/DevicePropertiesRequest.cs]: I used it because it is the comparison type used in the previous implementation that was proven to work:
https://github.com/Azure/device-simulation-dotnet/blob/0d37f326717db065dab767105763d29a0ece24bd/SimulationAgent/Simulation/DeviceStatusLogic/UpdateReportedProperties.cs#L160

However, ToString is overall not very efficient for comparison because every property's value is accessed. I thought of possibly implementing a comparer but because the objects are generic, couldn't think of a better way to go about it. 

Any alternate suggestions?

File: Services/DevicePropertiesRequest.cs (hunk 6 above)
Comment (jillcary) [reply Services/DevicePropertiesRequest.cs]: The implementation calls dictionary.Add which can throw an ArgumentNull exception

File: Services/DevicePropertiesRequest.cs (hunk 10 above)
Comment (dluc) [reply Services/DevicePropertiesRequest.cs]: I think that "every value is accessed" is a requirement in order to compare the value, right?

In this before-vs-after example, to detect that "five" was added under the "someproperty" property,
the code needs to see all the props:

```
"desired-properties": {
	"someproperty": {
	    "one": {
	       "two": 2,
	       "three": 3
	    },
	    "four": [ 0,1,2,3 ]
	}
}
```

```
"desired-properties": {
	"someproperty": {
	    "one": {
	       "two": 2,
	       "three": 3,
	       "five": 5
	    },
	    "four": [ 0,1,2,3 ]
	}
}
```

Comment (jillcary) [reply Services/DevicePropertiesRequest.cs]: Since this meets the parity requirements for RMv2 Preview, I've filed a bug here that we can address:
https://github.com/Azure/device-simulation-dotnet/issues/194
